inst/tests/tests.Rraw

```diff
@@ -12468,6 +12468,43 @@ gs = groupingsets(d, j = sum(val), by = c("a", "b", "c"),
                            character()), id=TRUE)
 test(1961, cb, gs)
 
+# #3185 -- .SDcols = integer(0L) completes gracefully
+DT = data.table(a = 1:10)
+test(1962.1, DT[ , .SD, .SDcols = integer(0L)], data.table(NULL))
+test(1962.2, DT[ , .SD, .SDcols = character(0L)], data.table(NULL))
+
+# #1878 -- patterns API in .SDcols
+library(data.table)
+DT = data.table(
+  i = 1:10,
+  c = c("a", "b", "c", "d", "e", "f", "g", "h", "i", "j"),
+  V1 = c(0.4, -0.1, -1.1, -2.6, -0.1, -1.3, 0.3, -2.1, -0.6, 0.9),
+  V2 = c(-0.1, -2.5, -1, -0.1, -0.5, -0.7, -1, -2.1, 2.7, -1.2),
+  V3 = c(1.1, -1.6, 0.7, 1.6, -1.4, 1, -0.6, 1.2, -0.8, 0.1),
+  V4 = c(1.3, -0.8, 2.3, -0.7, 0.5, 0.5, 0.2, 0.7, -1.4, 0.8),
+  V5 = c(-0.1, -0.5, 1.5, -0.5, 1.9, 0.2, -0.1, -0.7, -1.7, -0.9),
+  V6 = c(0.8, -1.3, -0.7, -0.3, 1.4, 0.7, 0.4, 0.3, -1.6, -1.3),
+  V7 = c(-0.1, 0.8, 0.7, -0.2, -2, 0.5, 0.4, -0.2, -1.2, -0.7),
+  V8 = c(0.7, -1, 1.3, 0.5, 0.2, 0.8, 0.6, -1.4, -2, -0.1),
+  V9 = c(0.2, -0.1, 1.2, -0.5, 1.4, 1, 0.2, 0.7, 0.4, 1.6),
+  V10 = c(0.8, 0.7, -1.2, -0.9, -0.6, 0.4, -2.3, 2.2, 0.5, -1.4)
+)
+
+test(1963.1, DT[ , lapply(.SD, sum), .SDcols = patterns('^V')],
+     data.table(V1 = -6.3, V2 = -6.5, V3 = 1.3, V4 = 3.4, V5 = -0.9,
+                V6 = -1.6, V7 = -2, V8 = -0.4, V9 = 6.1, V10 = -1.8))
+# multiple pattens --> intersection of patterns
+test(1963.2, DT[ , lapply(.SD, sum), .SDcols = patterns('^V[0-4]', '^V[5-9]')],
+     data.table(NULL))
+test(1963.3, DT[ , lapply(.SD, sum), .SDcols = patterns('^V[02468]', '^V[48]')],
+     data.table(V4 = 3.4, V8 = -0.4))
+
+# also with !/- inversion
+test(1963.4, DT[ , lapply(.SD, sum), .SDcols = !patterns('^c|i')],
```

Comment:
you mentioned `!/-` but tested only `!`

Reply:
As the two cases are literally handled side-by-side, there's not as of now any way for the behavior to differ for the two approaches? Is why I only tested the one:

https://github.com/Rdatatable/data.table/blob/87a7bb6b5ddefd9602733957ead50b1fa13f44a2/R/data.table.R#L1009

Reply:
I'm kicking the can to downstream... if someone wants to treat them differently at some point, add the tests then 😛 

The only worry is some upstream change that somehow changes how `-` and `!` work in R in general? But I think as long as lazy eval remains as is and they both remain functions we'll be fine.